minikube-iso: add cri-o runtime

vbatts opened the pull request:
https://github.com/kubernetes-incubator/cri-o

Updated the runc version to its latest master commit.
Got crio into the automounter to get off the tmpfs

This feature bubbles up to the minikube command by reusing the
`--container-runtime=` flag, by enabling the value of "`crio`" (`minikube start --container-runtime=crio`), while
the flags/config passed to localkube are more like k8s
(`--container-runtime=remote --remote-runtime-endpoint=/var/run/crio.sock`)

This is mostly ready for review. It is still lacking having
--insecure-registry plumbed through, but for now the policy.json is
open.

Will rebase on crio master once https://github.com/kubernetes-incubator/cri-o/pull/940 is merged.

Signed-off-by: Vincent Batts <vbatts@hashbangbash.com>

r2d4 commented:
@minikube-bot ok to test

vbatts commented:
sorry. I had to update the checksum. Please retest.

r2d4 commented:
@minikube-bot retest this please

r2d4 commented:
@minikube-bot retest this please

r2d4 commented:
@minikube-bot retest this please

vbatts commented:
interesting that the cross build fails on a `mkdir`, i'll look into why that fails

vbatts commented:
updated and rebased

vbatts commented:
@r2d4 once I get the dust settled for this addition, a next step will be adding logic to the provisioner about not starting services that are unneeded for particular `--container-runtime=`. Opinions about that?

r2d4 commented:
@vbatts I'm guessing you mean the docker service primarily?  I think that should be fine and I'm open to whatever changes we need to make, although you might encounter some difficultly with the leaky abstractions in that a few things might wait on the docker endpoint to be available.

We're also planning to hopefully move away from localkube in the future and use kubeadm, for which we already have the `--bootstrapper kubeadm` flag for.  I plan on sending a PR soon to make sure that all the configuration we set for localkube gets passed through kubeadm similarly.  I can make sure this works for that bootstrapper, but just something to keep in mind.

vbatts commented:
@r2d4 i expect leaky abstractions to be going away. But while testing this, it is both docker and rkt services are started.

vbatts commented:
ohman go1.7. Let me work on getting go1.8 into the build :pray:

r2d4 commented:
Ah, actually we have a dockerized build, if you add this

https://github.com/kubernetes/minikube/blob/master/hack/jenkins/minikube_cross_build_and_upload.sh#L38
This line should be changed to:
`IN_DOCKER=1 make out/minikube.iso`

r2d4 commented:
Jenkins just calls that script for the cross build job

vbatts commented:
ah. lemme give it a try

vbatts commented:
while that will show that it builds in such an environment, that line change would affect the ISO release process, no?

r2d4 commented:
Ah, you're right.  I thought our release process built the ISO in docker also, but it doesn't.  The makefile rules around the ISO could use some refactoring.  

```diff
diff --git a/Makefile b/Makefile
index a3f7e48b4..3abad1a3e 100755
--- a/Makefile
+++ b/Makefile
@@ -288,7 +288,7 @@ $(ISO_BUILD_IMAGE): deploy/iso/minikube-iso/Dockerfile
        @echo "$(@) successfully built"
 
 .PHONY: release-iso
-release-iso: minikube_iso checksum
+release-iso: out/minikube.iso checksum
        gsutil cp out/minikube.iso gs://$(ISO_BUCKET)/minikube-$(ISO_VERSION).iso
        gsutil cp out/minikube.iso.sha256 gs://$(ISO_BUCKET)/minikube-$(ISO_VERSION).iso.sha256
 
diff --git a/hack/jenkins/build_iso.sh b/hack/jenkins/build_iso.sh
index 32cc67131..6d76be548 100644
--- a/hack/jenkins/build_iso.sh
+++ b/hack/jenkins/build_iso.sh
@@ -20,4 +20,4 @@
 #      ISO_VERSION = the suffix for the iso (i.e. minikube-$(ISO_VERSION).iso)
 
 set -e
-${ARGS} make release-iso
+${ARGS} IN_DOCKER=1 make release-iso
```

That patch should fix the release job also

vbatts commented:
hmm. Still appears to be failing due to go1.7 compiler

r2d4 commented:
Ok, suspiciously there isn't even a go tool chain in the docker image we use to build the iso.  I'm not sure why we use a different image to build the iso, and I don't think that image is being build everytime.  

I'll just go ahead and update the build slave to go 1.8, since that seems like the path of least resistance here.  I'll update this issue when its upgraded.

r2d4 commented:
upgraded

@minikube-bot retest this please

vbatts commented:
I'll remove that docker variable when I get to my desk

On Tue, Sep 26, 2017, 13:37 Matt Rickard <notifications@github.com> wrote:

> upgraded
>
> @minikube-bot <https://github.com/minikube-bot> retest this please
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/minikube/pull/1998#issuecomment-332276291>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEF6VshL2C3CVtOTY-0dXr9BvACfzAEks5smTYmgaJpZM4Pg4vc>
> .
>

vbatts commented:
i've removed that `IN_DOCKER=1 make out/minikube.iso` commit.

Also, with https://github.com/kubernetes-incubator/cri-o/pull/940 merged, I've switched back to cri-o master commit (not pointing to my branch).

Not sure why the windows environment failed, but we'll see if it works this next time.

r2d4 commented:
the windows machine is in a bad state and we've had some problems with it in the past, since the tests passed on most of the other platforms, I think it should be fine.

r2d4 commented:
After this is merged I'll create an integration test using the `--container-runtime=crio` flag, since theres a little bit of setup on our side for adding a new integration test.

Pulled it down locally and ran it, seems to be working fine.

vbatts commented:
cool. Like I said, the --insecure-registry still needs to be plumbed
through. And wouldn't mind getting the provisioner logic to result in fewer
services. Those can be done in future PRs.

On Wed, Sep 27, 2017 at 12:51 PM Gerard Braad <notifications@github.com>
wrote:

> *@gbraad* approved this pull request.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/minikube/pull/1998#pullrequestreview-65612543>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEF6WqnkTOpuwq4DIkbV8Sl03BRnUSlks5smn0XgaJpZM4Pg4vc>
> .
>

r2d4 commented:
Thanks for working through this!

vbatts commented:
@r2d4 when I run:
```shell
kubectl run httpd --image=httpd
kubectl describe pod httpd
```

I see 
```
[...]
    Container ID:       runc://1d753ae46099f0a2dca86b1d658d913bcabbc9d48c96b4d2bd56b8857481c413
[...]
```

vbatts commented:
oh sorry, you said describe node, yes i see that same thing. I'll check into that.

rhatdan commented:
:+1:

vbatts commented on deploy/iso/minikube-iso/package/crio-bin/crio.conf:
```diff
@@ -0,0 +1,149 @@
+image_volumes = "mkdir"
+
+# insecure_registries is used to skip TLS verification when pulling images.
+insecure_registries = [
```
once i saw that there were dynamic service files generated, I thought the same thing.
